### PR TITLE
Fixes cache image with incorrect OS on windows

### PR DESCRIPTION
**Pull request by micahyoung**
This uses and updated version of `imgutil` to create cache images based on the `runtime.GOOS` of the lifecycle executable.

Cache images will then be created or with the correct `os` to the value which triggers all subsequent image operations to build appropriate layers for the image. If a previous image was created it will be re-used with the existing format - only new images will be initialized with corrected OS.

The latest version of `imgutil` also secures the test registries so it can only be written when they have credentials.

Fixes issue 529:

> Pack allows users to create cache images. When creating a cache image from a WCOW machine, I would expect those cache images to themselves be WCOW images, but they are instead linux containers.
>
> > When creating a new cache image, I think this gets called… so the cache will default to Linux even if running on Windows
>
> https://github.com/buildpacks/imgutil/blob/30601e371ce385a990992de9ed9ff7b373f7c524/remote/remote.go#L108

Signed-off-by: Micah Young <ymicah@vmware.com>

### CODE REVIEW

File: cache/image_cache.go [hunk 1]
```diff
@@ -46,6 +47,10 @@ func NewImageCacheFromName(name string, keychain authn.Keychain) (*ImageCache, e
 	if err != nil {
 		return nil, fmt.Errorf("creating new cache image %q: %v", name, err)
 	}
+	if err := emptyImage.SetOS(runtime.GOOS); err != nil {
+		return nil, fmt.Errorf("setting OS new cache image %q: %w", name, err)
+	}
```
Comment (ekcasey) [inline cache/image_cache.go:52]: I am going to assume we need to do something like this as well?

https://github.com/buildpacks/pack/blob/main/internal/buildpackage/builder.go#L109

Comment (micahyoung) [reply cache/image_cache.go:52]: Yeah, exactly. I was just checking to see if this had any adverse side effects in CI, since I wasn't seeing any failures locally, despite the images being invalid.

**Comment by micahyoung** [comment]
With these two commits, it appears to be generating valid Windows cache images now:

1. Clone and build sample using `cnbs/sample-builder:dotnet-framework-1809` (rebuilt with PR lifecycle at `micahyoung/sample-builder-fix-cache-image:dotnet-framework-1809`)

```bash
git clone https://github.com/microsoft/dotnet-framework-docker

pack build \
  micahyoung/aspnetapp:latest \
  --path dotnet-framework-docker/samples/aspnetapp \
  --publish --cache-image micahyoung/aspnetapp-cache:latest \
  --builder micahyoung/sample-builder-fix-cache-image:dotnet-framework-1809 \
  --trust-builder
```

Output 
```
===> DETECTING
samples/dotnet-framework 0.0.1
===> ANALYZING
Previous image with name "micahyoung/aspnetapp:latest" not found
Layer cache not found
===> RESTORING
Layer cache not found
===> BUILDING
<redacted for brevity - lots of nuget fetching>
===> EXPORTING
Adding 1/1 app layer(s)
Adding layer 'launcher'
Adding layer 'config'
Adding layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
*** Images (sha256:8a7076ad55f6c093f5a38cfb17ec999c7b99f3f9e533b9b7f391bc61fe478e3c):
      micahyoung/aspnetapp:latest
Layer cache not found
Adding cache layer 'samples/dotnet-framework:dotnet-framework'
Successfully built image micahyoung/aspnetapp:latest
```


2. pack build ^^^ again
Output:
```
===> DETECTING
samples/dotnet-framework 0.0.1
===> ANALYZING
Restoring metadata for "samples/dotnet-framework:dotnet-framework" from cache
===> RESTORING
Restoring data for "samples/dotnet-framework:dotnet-framework" from cache
===> BUILDING
<redacted for brevity - less than before, no nuget fetching>
===> EXPORTING
Reusing layers from image 'index.docker.io/micahyoung/aspnetapp@sha256:8a7076ad55f6c093f5a38cfb17ec999c7b99f3f9e533b9b7f391bc61fe478e3c'
Adding 1/1 app layer(s)
Reusing layer 'launcher'
Reusing layer 'config'
Reusing layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
*** Images (sha256:3473e8d88231526c36626de05c570cb0e11b6ceadc6e6afa63cde0bff87f4637):
      micahyoung/aspnetapp:latest
Reusing cache layer 'samples/dotnet-framework:dotnet-framework'
Successfully built image micahyoung/aspnetapp:latest
```

3. Docker pulls successfully
```bash
docker pull micahyoung/aspnetapp-cache:latest
```
Output:
```
latest: Pulling from micahyoung/aspnetapp-cache
8075243551c3: Already exists 
704d7bf6878f: Pull complete 
Digest: sha256:e00bd2337acf441c7d21a1d1b4febe5a04f9a0b1559b79513127a44cfc352704
Status: Downloaded newer image for micahyoung/aspnetapp-cache:latest
docker.io/micahyoung/aspnetapp-cache:latest
```
4. Crane config shows correct metadata + 2 layers
```bash
crane config micahyoung/aspnetapp-cache:latest | jq
```
Output:
```
{
  "architecture": "amd64",
  "created": "1980-01-01T00:00:01Z",
  "history": [
    {
      "created": "1980-01-01T00:00:01Z"
    },
    {
      "created": "1980-01-01T00:00:01Z"
    }
  ],
  "os": "windows",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:067833dadeca9180bb71a211248cbc0f6ada05499bfdef07dfb04e2eb96e7c82",
      "sha256:3c39707d8bcc5ed845f1dd0024d4cb85d4c7dc673b31005fc2283e2172bcd524"
    ]
  },
  "config": {
    "Labels": {
      "io.buildpacks.lifecycle.cache.metadata": "{\"buildpacks\":[{\"key\":\"samples/dotnet-framework\",\"version\":\"0.0.1\",\"layers\":{\"dotnet-framework\":{\"sha\":\"sha256:3c39707d8bcc5ed845f1dd0024d4cb85d4c7dc673b31005fc2283e2172bcd524\",\"data\":null,\"build\":true,\"launch\":false,\"cache\":true}}}]}"
    }
  }
}
```

5. Dive aggregated view shows both layers and correct cache layer contents + shim layer contents
```bash
dive micahyoung/aspnetapp-cache:latest
```
Output: (`dotnet-framework` has many subdirectories, collapsed to show shim layer contents)
```
│ Layers ├──────────────────────────────────────── ┃ ● Aggregated Layer Contents ┣━━━━━━━━━━━━━━━━━━━
Cmp   Size  Command                                ├── Files                                         
     12 kB  FROM 8a4dd6e94861fde                   │   ├── Windows                                   
     50 MB                                         │   │   └── System32                              
                                                   │   │       └── config                            
│ Layer Details ├───────────────────────────────── │   │           ├── DEFAULT                       
                                                   │   │           ├── SAM                           
Tags:   (unavailable)                              │   │           ├── SECURITY                      
Id:     2e78deed7f3e0ed63839ca9f30b5543040f6af7dac │   │           ├── SOFTWARE                      
7939dd488f7854b7f82956                             │   │           └── SYSTEM                        
Digest: sha256:3c39707d8bcc5ed845f1dd0024d4cb85d4c │   └── layers                                    
7dc673b31005fc2283e2172bcd524                      │       └── samples_dotnet-framework              
Command:                                           │           └─⊕ dotnet-framework                  
                                                   ├── Hives                                         
                                                   └── UtilityVM                                     
│ Image Details ├─────────────────────────────────     └── Files                                     
                                                           └── EFI                                   
                                                               └── Microsoft                         
Total Image size: 50 MB                                            └── Boot                          
Potential wasted space: 0 B                                            └── BCD                       
Image efficiency score: 100 %                                                                        
                                                                                                     
Count   Total Space  Path                                                                            
```

File: cache/image_cache.go (hunk 1 above)
Comment (micahyoung) [reply cache/image_cache.go:72]: I've now added the shim layer and CI is still passing. I intend to rework this a bit more add more testing (also try to re-enable Windows acceptance tests that use this). 

I'm still not sure whether adding the shim-layer should be moved into `imgutil`. The `image.AddLayer(tarPath)` interface kinda goes against adding layers any other way, plus one of `pack`'s `WorkableImage` implementations doesn't use an `imgutil` image so would still need to add the shim layer anyways.

Did you have any thoughts on a cleaner interface for now?

Comment (ekcasey) [reply cache/image_cache.go:72]: 1. I think an `AddLayerFromReader` would make sense as an imgutil feature.
1. It would also make sense in `imgutil` added the shim layer when creating an empty windows image.
1. I don't really understand why we added an implementation for image layouts directly to pack instead of adding it to `imgutil`, but I don't have all the context on that one.

File: cache/image_cache.go [hunk 2]
```diff
@@ -17,23 +21,35 @@ import (
...
+	newImage          imgutil.Image
+	generatedLayerDir string
 }
 
-func NewImageCache(origImage imgutil.Image, newImage imgutil.Image) *ImageCache {
-	return &ImageCache{
-		origImage: origImage,
-		newImage:  newImage,
+func NewImageCache(origImage imgutil.Image, newImage imgutil.Image) (*ImageCache, error) {
+	generatedLayerDir, err := ioutil.TempDir("", "imace-cache-generated-layers")
```
Comment (natalieparellano) [inline cache/image_cache.go:31]:
```suggestion
	generatedLayerDir, err := ioutil.TempDir("", "image-cache-generated-layers")
```

File: go.mod [hunk 3]
```diff
@@ -1,5 +1,7 @@
 module github.com/buildpacks/lifecycle
 
+replace github.com/buildpacks/imgutil => github.com/micahyoung/buildpacks-imgutil v0.0.0-20210224123626-3a8e49030b47
```
Comment (natalieparellano) [inline go.mod:3]: Does this depend on a change to imgutil?

Comment (natalieparellano) [reply go.mod:3]: Ohhh I see - `WithPlatform`

**Comment by natalieparellano** [comment]
@micahyoung could we undraft this one now that https://github.com/buildpacks/imgutil/pull/95 is merged?

**Comment by micahyoung** [comment]
@natalieparellano Thanks, yeah, I'll do that right now. I'm separately still fixing conflicts in the analyzer test PR https://github.com/buildpacks/lifecycle/pull/541, so that will come a little bit later.

File: cache/image_cache.go [hunk 4]
```diff
@@ -22,18 +23,19 @@ type ImageCache struct {
 	newImage  imgutil.Image
 }
 
-func NewImageCache(origImage imgutil.Image, newImage imgutil.Image) *ImageCache {
+func NewImageCache(origImage imgutil.Image, newImage imgutil.Image) (*ImageCache, error) {
 	return &ImageCache{
 		origImage: origImage,
 		newImage:  newImage,
-	}
+	}, nil
```
Comment (natalieparellano) [inline cache/image_cache.go:30]: Why is this needed? Couldn't we just keep doing `return NewImageCache(origImage, emptyImage), nil`?

Comment (micahyoung) [reply cache/image_cache.go:30]: Oh, good catch. I'll fix this. I removed the section that threw the error and forgot to clean this up.

**Comment by natalieparellano** [comment]
Fixes #529
